native/src/menu.rs

```diff
@@ -0,0 +1,81 @@
+//! Build menus for your application.
+use crate::keyboard::Hotkey;
+
+/// Menu representation.
+///
+/// This can be used by `shell` implementations to create a menu.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Menu<Message> {
+    items: Vec<MenuEntry<Message>>,
+}
+
+impl<Message> Menu<Message> {
+    /// Creates an empty [`Menu`].
+    pub fn new() -> Self {
+        Menu { items: Vec::new() }
+    }
+
+    /// Adds an item to the [`Menu`].
+    pub fn item<S: Into<String>>(
+        mut self,
+        content: S,
+        hotkey: impl Into<Option<Hotkey>>,
+        on_activation: Message,
+    ) -> Self {
+        let content = content.into();
+        let hotkey = hotkey.into();
+
+        self.items.push(MenuEntry::Item {
+            on_activation,
+            content,
+            hotkey,
+        });
+        self
+    }
+
+    /// Adds a separator to the [`Menu`].
+    pub fn separator(mut self) -> Self {
+        self.items.push(MenuEntry::Separator);
+        self
+    }
+
+    /// Adds a dropdown to the [`Menu`].
+    pub fn dropdown<S: Into<String>>(
+        mut self,
+        content: S,
+        submenu: Menu<Message>,
+    ) -> Self {
+        let content = content.into();
+
+        self.items.push(MenuEntry::Dropdown { content, submenu });
+        self
+    }
+
+    /// Returns a [`MenuEntry`] iterator.
+    pub fn iter(self) -> std::vec::IntoIter<MenuEntry<Message>> {
```

Comment:
Ideally I don't want to expose this method and `MenuEntry`, allowing the end-user to only use our constructor methods (`new`, `item`, `dropdown`, `separator`), but then how would the shell implementations `iter` through the `Menu` to actually build it? Maybe there's a better solution here?

Reply:
I think it's fine for `menu::Entry` to be completely public. I don't see what kinds of guarantees we would be capturing with opacity here. The methods can be seen as just convenient helpers with some generics.